### PR TITLE
Added “SemanticModel.WithSuppressAccessChecks” extension method.

### LINKED ISSUE #378: Add SemanticModel.WithoutAccessChecks()
From: Omer Raviv 
Sent: Thursday, November 6, 2014 12:43 PM
Subject: RE: Roslyn issues
 
Hey Guys,
 
Thanks a lot for your help, Kevin.
 
Of the two issues, 1# is the most crucial to us and the one we just can't ship without, so we’d really appreciate any possible help you could give us!
 
If I understand correctly, what we're asking for here amount to:
1. Add a public API on `SemanticModel` to request a new `SemanticModel` that ignores accessibility checks, something like `SemanticModel.WithSuppressAccessibilityChecks`.
2. Change every line of code where a `Binder` is created with `BindingOptions.SemanticModel` (9 places by my count), so that it also includes `BindingFlags.SuppressAccessibilityChecks` if that's what the `SemanticModel` specifies.
3. Change the line in `Binder.IsSymbolAccessibleConditional` that says `throw ExceptionUtilities.Unreachable;` to say `return true;`
 
I'm sure there are many API design considerations here that I'm not aware of, but this is really all we're asking. Without this, we can’t get out of Roslyn the behavior that OzCode inherently depends on, and we’re kind of stuck.
 
Thanks a lot!
- Omer
 
-------------------- 
From: Tomas Matousek
Sent: Thursday, November 6, 2014 09:45 PM
Subject: RE: Roslyn issues
 
I think we need the same features for our EEs and to support interactive.
 
Re #1: The current implementation of EEs heavily depends on internals of the compiler. I’d like us to use less internal APIs and more public APIs that customers like CodeValue, Xamarin and others could leverage as well.
Re #2: To support interactive we are going to need a metadata reference whose internals are imported. Seems like exactly what Pavel is asking for. 
 
Not sure when this work will be done. We are currently stretched thin.
 
Tomas

--------------------- 
From: Kevin Pilch-Bisson 
Sent: Thursday, November 6, 2014 11:36 AM
Subject: RE: Roslyn issues
 
Hey Guys,
 
Chuck/Tomas – OzCode is a VS extension that adds a bunch of functionality to the debugging experience.  They are looking at moving over to using Roslyn, but there are a couple of issues blocking them at the moment.  I’m wondering if you have ideas for workarounds, or thoughts about how we do these things in our own EEs.
 
1.       Avoiding access checks so that they can use `SemanticModel` to look at private/internal data at debug time.  They tried changing the implementation of `IsSymbolAccessibleConditional` to just return `true`, and that works for them, but they don’t want to fork Roslyn and build their own compilation models just to get this working.
2.       They’d like to import internal metadata, but `CompilationOptions.WithMetadataImportOptions` is internal.  Even if it wasn’t, they would need some way to tell VS to use that, so that they could re-use our compilations.
 
Any ideas?
--
Kevin Pilch-Bisson
    http://twitter.com/Pilchie

[issue ported from TFS DevDiv 1078951]

### PR DESCRIPTION
**Pull request by OmerRaviv**
The `WithSupressAccessChecks` method creates a SemanticModel that ignores accessibility rules when answering semantic questions.

This allows consumers to ask semantic questions using the same semantic rules as the ones used in debugger’s Expression Evaluator, where accessibility rules do not matter. This API is an absolute necessity for 3rd parties (such as OzCode) who want to create debugger-related productivity tools on top of Roslyn.

Added unit tests for C# and VB.NET to cover both regular and speculative analysis.

fixes #378

### CODE REVIEW

File: src/Compilers/Core/Portable/Compilation/SemanticModel.cs [hunk 1]
```diff
@@ -63,6 +63,16 @@ public SyntaxTree SyntaxTree
         protected abstract SyntaxTree SyntaxTreeCore { get; }
 
         /// <summary>
+        /// Returns true if this is a SemanticModel that ignores accessibility rules when answering semantic questions.
+        /// </summary>
+        public bool IsSuppressingAccessChecks { get; protected set; }
```
Comment (gafter) [inline src/Compilers/Core/Portable/Compilation/SemanticModel.cs]: The setter should be internal.

File: src/Compilers/VisualBasic/Portable/Compilation/SemanticModel.vb [hunk 2]
```diff
@@ -106,6 +106,14 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         ''' </summary>
         Friend MustOverride Function GetCrefReferenceSymbolInfo(crefReference As CrefReferenceSyntax, options As SymbolInfoOptions, Optional cancellationToken As CancellationToken = Nothing) As SymbolInfo
 
+        Public Overrides Function WithSuppressAccessChecks() As SemanticModel
+            Dim moduleSymbol = TryCast(Compilation.SourceModule, SourceModuleSymbol)
```
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Compilation/SemanticModel.vb]: If the current model is already ignoring accessibility, we should just return it.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 3]
```diff
@@ -51,6 +51,11 @@ internal abstract class CSharpSemanticModel : SemanticModel
         /// </summary>
         internal abstract CSharpSyntaxNode Root { get; }
 
+
+        public override SemanticModel WithSuppressAccessChecks()
+        {
+            return new SyntaxTreeSemanticModel(Compilation, this.SyntaxTree, suppressAccessChecks: true);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: If the current model is already suppressing the checks, we should just return it.

File: src/Compilers/Core/Portable/Compilation/SemanticModel.cs [hunk 4]
```diff
@@ -63,6 +63,16 @@ public SyntaxTree SyntaxTree
...
 
         /// <summary>
+        /// Returns true if this is a SemanticModel that ignores accessibility rules when answering semantic questions.
+        /// </summary>
+        public bool IsSuppressingAccessChecks { get; internal set; }
+
+        /// <summary>
+        /// Creates a SemanticModel that ignores accessibility rules when answering semantic questions.
+        /// </summary>
+        public abstract SemanticModel WithSuppressAccessChecks();
```
Comment (AnthonyDGreen) [inline src/Compilers/Core/Portable/Compilation/SemanticModel.cs]: I'd much rather expose this as a new optional parameter added to the existing Compilation.GetSemanticModel overloads which takes a Boolean flag for whether accessibility checks should be suppressed for the purpose of lookup which defaults to false. Perhaps bool ignoreAccessibility or suppressAccessChecks.

File: src/Compilers/Core/Portable/Compilation/SemanticModel.cs [hunk 5]
```diff
@@ -63,6 +63,16 @@ public SyntaxTree SyntaxTree
         protected abstract SyntaxTree SyntaxTreeCore { get; }
 
         /// <summary>
+        /// Returns true if this is a SemanticModel that ignores accessibility rules when answering semantic questions.
+        /// </summary>
+        public bool IsSuppressingAccessChecks { get; internal set; }
```
Comment (AnthonyDGreen) [inline src/Compilers/Core/Portable/Compilation/SemanticModel.cs]: HasAccessChecksSuppressed

**Comment by OmerRaviv** [comment]
Thanks a lot for the feedback, @AnthonyDGreen. That's indeed a much cleaner solution. Fixed.

**Comment by gafter** [comment]
Note that this fixes #26.

**Comment by gafter** [comment]
Note that this fixes #378

**Comment by AnthonyDGreen** [comment]
@tmat, @KevinH-MS, @cston, @ManishJayaswal, does this aid or address any debugging/REPL scenarios that also require bypassing access checks in the SemanticModel?

**Comment by mattwar** [comment]
A debugging scenario is freer with accessibility than the REPL scenarios we've discussed. One REPL scenario discussed was to have an option that allowed the REPL or script code to access internals as if internals-visible-to had been set. Note, a REPL might eventually be used in a debugging scenario, so this could come in to play for that.

**Comment by gafter** [comment]
@OmerRaviv Can you please "sign" a contributor's agreement starting at https://cla.dotnetfoundation.org/

**Comment by OmerRaviv** [comment]
@gafter I've signed the dotnetfoundation CLA on the 18th of January using the email address that's associated with my GitHub account, omerraviv@gmail.com, and got the approval the next day.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 6]
```diff
@@ -2747,6 +2748,13 @@ internal Conversion ClassifyConversionForCast(int position, ExpressionSyntax exp
         /// <param name="typeParameter"></param>
         public abstract ITypeParameterSymbol GetDeclaredSymbol(TypeParameterSyntax typeParameter, CancellationToken cancellationToken = default(CancellationToken));
 
+        protected BinderFlags GetSemanticModelBinderFlags()
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: This should be internal, not protected, as it should not be part of the publicly visible API.

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 7]
```diff
@@ -29,10 +29,12 @@ internal partial class SyntaxTreeSemanticModel : CSharpSemanticModel
 
         private static readonly Func<CSharpSyntaxNode, bool> isMemberDeclarationFunction = IsMemberDeclaration;
 
-        internal SyntaxTreeSemanticModel(CSharpCompilation compilation, SyntaxTree syntaxTree)
+        internal SyntaxTreeSemanticModel(CSharpCompilation compilation, SyntaxTree syntaxTree, bool suppressAccessChecks = false)
         {
             this.compilation = compilation;
             this.syntaxTree = syntaxTree;
+            this.HasAccessChecksSuppressed = suppressAccessChecks;
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs]: Please use the pattern as for other things; make this a local field in only those subclasses that need it, and make the original property a virtual (you can make the base implemenetation return false).

File: src/Compilers/VisualBasic/Portable/Binding/SemanticModelBinder.vb [hunk 8]
```diff
@@ -8,15 +8,27 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+        Protected Sub New(containingBinder As Binder, Optional ignoreAccessibility As Boolean = False)
             MyBase.New(containingBinder)
+
+            _ignoreAccessibility = ignoreAccessibility
         End Sub
 
-        Public Shared Function Mark(binder As Binder) As Binder
+        Public Shared Function Mark(binder As Binder, Optional ignoreAccessibility As Boolean = False) As Binder
             Return If(
                 binder.IsSemanticModelBinder,
```
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Binding/SemanticModelBinder.vb]: This condition is insufficient to ensure the existing binder is right. You also need to test its accessibility ignorance.

**Comment by gafter** [comment]
This pull request failed its automatic merge and test. Please re-merge with the latest and resubmit the pull request.

**Comment by OmerRaviv** [comment]
@gafter I've addressed your latest comments and resubmitted the Pull Request as one squashed commit. I've updated the commit message to reflect the new design decisions.

**Comment by gafter** [comment]
@tmat Tomas, can you please review this? Do you think this API change makes sense?

**Comment by gafter** [comment]
:+1: This looks good to me.

@jaredpar @mattwar @AlekseyTs @VSadov Can you please check this too before it is pulled?

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 9]
```diff
@@ -2805,7 +2813,7 @@ public ILocalSymbol GetDeclaredSymbol(CatchDeclarationSyntax catchDeclaration, C
                     return null;
                 }
 
-                catchBinder = catchBinder.WithAdditionalFlags(BinderFlags.SemanticModel);
+                catchBinder = enclosingBinder.GetBinder(catchClause).WithAdditionalFlags(GetSemanticModelBinderFlags());
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: catchBinder = catchBinder.WithAdditionalFlags(GetSemanticModelBinderFlags());

**Comment by tmat** [comment]
Looks reasonable.

@rchande @amcasey Could we use this in our EE completion provider? The EE allows evaluating internal members but completion doesn't show them, which is a bummer since the IDE actively fights you when you try to write them.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 10]
```diff
@@ -51,6 +51,7 @@ internal abstract class CSharpSemanticModel : SemanticModel
         /// </summary>
         internal abstract CSharpSyntaxNode Root { get; }
 
+
         // Is this node one that could be successfully interrogated by GetSymbolInfo/GetTypeInfo/GetMemberGroup/GetConstantValue?
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Nit: double blank line.

File: src/Compilers/VisualBasic/Portable/Binding/AttributeSemanticModel.vb [hunk 11]
```diff
@@ -11,15 +11,15 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
     Friend NotInheritable Class AttributeSemanticModel
         Inherits MemberSemanticModel
 
-        Private Sub New(root As VisualBasicSyntaxNode, binder As Binder, Optional parentSemanticModelOpt As SyntaxTreeSemanticModel = Nothing, Optional speculatedPosition As Integer = 0)
-            MyBase.New(root, binder, parentSemanticModelOpt, speculatedPosition)
+        Private Sub New(root As VisualBasicSyntaxNode, binder As Binder, Optional parentSemanticModelOpt As SyntaxTreeSemanticModel = Nothing, Optional speculatedPosition As Integer = 0, Optional isSupressingAccessChecks As Boolean = False)
+            MyBase.New(root, binder, parentSemanticModelOpt, speculatedPosition, isSupressingAccessChecks)
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Binding/AttributeSemanticModel.vb]: Consider 'suppressAccessChecks' instead for consistency with C#.

**Comment by rchande** [comment]
@tmat Since it's just a public API on Compilation, we could have completion use this during debug time. Does this automatically work with Speculative SemanticModels?

File: src/Compilers/Core/Portable/Compilation/SemanticModel.cs [hunk 12]
```diff
@@ -63,6 +63,14 @@ public SyntaxTree SyntaxTree
         protected abstract SyntaxTree SyntaxTreeCore { get; }
 
         /// <summary>
+        /// Returns true if this is a SemanticModel that ignores accessibility rules when answering semantic questions.
+        /// </summary>
+        public virtual bool HasAccessChecksSuppressed
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Compilation/SemanticModel.cs]: Can we name this `SuppressAccessChecks`?  

The pattern of `SuppressY` is used a number of times in our code base both internally and publically.  The pattern of `HasYSuppressed` does not appear anywhere I could find though we do have a few occurrences of `IsYSuppressed` but they are all methods and internal.  

That pattern seems to be more consistent with other public, and even internal, APIs that we expose.  


@gafter what do you think?

**Comment by jaredpar** [comment]
:+1: 

I had one bit of feedback about the choice of API name here.  I think `SuppressAccessChecks` is more consistent with existing patterns `HasAccessChecksSuppressed`.  Wanted to get others feedback on that before merging.  

Other than that though it looks great!

**Comment by gafter** [comment]
SuppressAccessChecks  sounds a bit more like a command than a test.

@AnthonyDGreen  what do you think?

File: src/Compilers/Core/Portable/Compilation/SemanticModel.cs (hunk 12 above)
Comment (gafter) [inline src/Compilers/Core/Portable/Compilation/SemanticModel.cs]: What about IgnoresAccessibility?

**Comment by jaredpar** [comment]
Merging a couple of comment threads here.  The two alternate proposals for name are:

- SuppressAccessChecks
- IgnoresAccessibility

I am fine with either.  Both have precedent within the Roslyn code base and fit into existing patterns which was my primary concern. 

Small nit is I would use `IgnoreAccessibility` over `IgnoresAccessibility`.  There are no cases I could find of a plural form of ignores but plenty where the singular variety is used.

**Comment by OmerRaviv** [comment]
@rchande Yes, this automatically supports Speculative SemanticModels, see unit tests for proof :) 
@rchande @tmat @amcasey Yes, I think it would make sense to combine this functionality with #423, so that you can get IntelliSense to include inaccessible members while debugging. 

@jaredpar @gafter @AnthonyDGreen  I don't really have an opinion regarding naming either way, please let me know once you've reached consensus.

**Comment by AnthonyDGreen** [comment]
@jaredpar, I don't think we have that many Boolean flags in the API but in the past I've used "the if test". How does checking the flag look in an if:

```C#
if (semanticModel.HasAccessChecksSuppressed) { .. }
if (semanticModel.IgnoresAccessibility) { ... }
if (semanticModel.SuppressAccessChecks) { ... }
```
I think either of the first two read better in that context. This is also what leads me to prefer 'ignores' over 'ignore'. We had a similar discussiong when we named ControlFlowAnalysis/DataFlowAnalysis.Succeeded vs Success. If we're applying this inconsistently we should review the various cases holistically.

@terrajobst for weigh in on naming.

**Comment by jaredpar** [comment]
@AnthonyDGreen the choice of `Ignores` vs. `Ignore` doesn't really matter to me.  All I care about is consistency in the choices.  Today we do have APIs in the singular so if we make this plural I prefer we make all of the other plural as well (even if the APIs are internal).  The same is true for `HasYSuppressed`, it breaks an existing established pattern.

**Comment by gafter** [comment]
I can't find examples where we use the "singular" except in the imperative sense (of instructing the API to ignore something), such as LookupOptions.IgnoreAccessibility. This property test the setting.

**Comment by jaredpar** [comment]
Here are the cases I could find.  

Singular:

- AbstractRegionDataFlowPass.IgnoreOutSemantics
- DataFlowPass.IgnoreOutSemantics
- UnassignedVariablesWarker.IgnoreOutSemantics
- IgnoreBaseClassesInLookup 
- CandidateAnalysisResult.IgnoreExtensionMethods
- ComparisonOptions.IgnoreCase
- ComparisonOptions.IgnoreAssemblyKey
- SuppressOptions.IgnoreElastic

Plural:

- Binder.IgnoresAccessibility

**Comment by gafter** [comment]
The only ones of those that are public are options (commands) to ignore something.

**Comment by jaredpar** [comment]
Internal or not I think we should be consistent.

JaredPar from a phone
http://blog.paranoidcoding.com/
________________________________

**Comment by AnthonyDGreen** [comment]
They're in different contexts. One is a flag on SemanticModel describing a _property_ of the semantic model. The Ignores isn't plural, it's third person singular. I ignore, you ignore, he/she/it _ignores_. This is the same pattern we use for SemanticModel.IsSpeculativeSemanticModel method (or Object.Equals). I am, you are, he/she/it _is_.

This is a different grammatical context than the parameter which one passes in to GetSemanticModel to cause accessibility to be ignored - ignoreAccessibility. That's describing an _action_ that should be taken (sort of a second person command) and options follow that same convention.

The problem is that depending on where the code is written a piece of data will need to be specified or described in different persons. To the person acquiring the SemanticModel it's 2nd - ignoreAccessibility, within the SemanticModel it would be 1st - IgnoreAccessibility, and outside of the SemanticModel it will be in the 3rd - IgnoresAccessibility. Things get even muddier when properties are read-write because getting is usually in the 3rd and setting is usually in the 2nd. We can't even just standardize on one because - Object.Equals is 3rd person, and IsSpeculativeSemanticModel is 3rd person. I'm not even sure what the latter would be in the 1st or 2nd persons - BeSpeculative, Speculate, AmSpeculative, or just Speculative?

So which name is used where depends heavily on who the primary consumer is. The parameters and options are primarily commands for someone else (e.g. the Compilation, the parser) and should be written in the 2nd person, and the properties on SemanticModel are primarily attributes that should be written in the 3rd person.

I agree that the internal cases should be made consistent if they are in a context where that reasoning applies.

**Comment by gafter** [comment]
I am going to accept this pull request as-is and then submit a separate change to rename the public member. Please say something now if that sounds like a bad plan.

**Comment by jaredpar** [comment]
Thanks fine by me.

**Comment by terrajobst** [comment]
Sorry to be super later but hey -- better late than never :-)

> @terrajobst for weigh in on naming.

I'm with Jared. Consistency is key. As far as imperative vs. descriptive style goes: we have both. It usually depends on context. In general, it's good to design booleans to make them read nice in `if` checks. I believe the ones discussed all satisfy this:

```csharp
if (semanticModel.HasAccessChecksSuppressed) { .. }
if (semanticModel.IgnoresAccessibility) { ... }
if (semanticModel.SuppressAccessChecks) { ... }
```

**Pull request merged**